File: InstaPass-iOS/InstaPass/LoginHelper/LoginHelper.swift
//
//  LoginHelper.swift
//  InstaPass
//
//  Created by 法好 on 2020/4/25.
//  Copyright © 2020 yuetsin. All rights reserved.
//

import Alamofire
import Alamofire_SwiftyJSON
import Foundation
import SwiftyJSON

class LoginHelper {
    // always assume login until one request gets unauthorized
    static var isLogin: Bool = true

    static func login(username: String, password: String, handler: @escaping (RequestResponse) -> Void) {
        let loginParams: Parameters = [
            "username": username,
            "password": password,
        ]
        RequestManager.request(type: .post,
                               feature: .login,
                               params: loginParams,
                               success: { jsonResp in
                                   LoginHelper.isLogin = true
                                   UserPrefInitializer.jwtToken = jsonResp["jwt_token"].stringValue
                                   let appDelegate = UIApplication.shared.delegate as? AppDelegate
                                   appDelegate?.sendJwtToken(token: UserPrefInitializer.jwtToken)
                                   handler(.ok)
                               }, failure: { _ in
                                   LoginHelper.isLogin = false
                                   handler(.unauthorized)
        })
    }

    static func logout(handler: (RequestResponse) -> Void) {
        isLogin = false
        UserPrefInitializer.jwtToken = ""
        
        let appDelegate = UIApplication.shared.delegate as? AppDelegate
        appDelegate?.sendJwtToken(token: "")
    }
}

File: InstaPass-Android/app/src/main/java/io/github/yuetsin/instapass/login/LoginHelper.kt
package io.github.yuetsin.instapass.login

import io.github.yuetsin.instapass.preferences.UserPrefInitializer.Companion.jwtToken
import io.github.yuetsin.instapass.request.FeatureTypeEnum
import io.github.yuetsin.instapass.request.RequestTypeEnum
import io.github.yuetsin.instapass.request.RequestsManager
import org.json.JSONObject

class LoginHelper {
    companion object {
        // always assume login until one request failed
        var isLogin = true

        fun login(
            username: String,
            password: String,
            success: () -> Void,
            failure: (String) -> Void
        ) {
            RequestsManager.request(RequestTypeEnum.Post, FeatureTypeEnum.Login, "", JSONObject().put("username", username).put("password", password),
                {
                    isLogin = true
                    jwtToken = it["jwt_token"].toString()
                    success()
                }, {
                    isLogin = false
                    failure(it)
                })
        }

        fun logout(success: () -> Void, failure: () -> Void) {
            isLogin = false
            jwtToken = ""

            success()
        }
    }
}